File: cpp/src/core/params.cpp
#include "nav_study/core/params.hpp"

#include <stdexcept>

#include "nav_study/core/yaml.hpp"

namespace nav_study::core {
namespace {

[[noreturn]] void fail(const std::string& msg) { throw std::runtime_error("param error: " + msg); }

bool is_integer_literal(const std::string& s) {
  if (s.empty()) return false;
  size_t i = (s[0] == '-' || s[0] == '+') ? 1 : 0;
  if (i >= s.size()) return false;
  for (; i < s.size(); ++i) {
    if (s[i] < '0' || s[i] > '9') return false;
  }
  return true;
}

void check_range(const std::string& name, double value, const std::optional<double>& lo,
                 const std::optional<double>& hi) {
  if (lo && value < *lo) fail("'" + name + "' below min");
  if (hi && value > *hi) fail("'" + name + "' above max");
}

}  // namespace

ParamSet ParamSet::from_yaml(const std::string& path) {
  YamlNode root = parse_yaml_file(path);
  if (!root.is_map()) fail("config root must be a mapping: " + path);

  ParamSet set;
  set.algorithm_ = root.at("algorithm").as_string();
  set.category_ = root.at("category").as_string();
  if (set.category_ != "global_planning" && set.category_ != "local_planning" &&
      set.category_ != "multi_agent") {
    fail("unknown category '" + set.category_ + "'");
  }

  const YamlNode& params = root.at("params");
  if (!params.is_seq() && !params.is_null()) fail("'params' must be a sequence");

  for (const YamlNode& decl : params.seq) {
    if (!decl.is_map()) fail("each param must be a mapping");
    std::string name = decl.at("name").as_string();
    std::string type = decl.at("type").as_string();
    const YamlNode& def = decl.at("default");
    decl.at("description");  // required by schema

    std::optional<double> lo, hi;
    if (decl.has("min")) lo = decl.at("min").as_double();
    if (decl.has("max")) hi = decl.at("max").as_double();

    if (type == "int") {
      if (!is_integer_literal(def.as_string())) fail("'" + name + "' default is not an integer");
      long long v = def.as_int();
      check_range(name, static_cast<double>(v), lo, hi);
      set.values_[name] = static_cast<int>(v);
    } else if (type == "float") {
      double v = def.as_double();
      check_range(name, v, lo, hi);
      set.values_[name] = v;
    } else if (type == "bool") {
      set.values_[name] = def.as_bool();
    } else if (type == "string") {
      set.values_[name] = def.as_string();
    } else if (type == "enum") {
      if (!decl.has("choices")) fail("'" + name + "' enum needs choices");
      std::string v = def.as_string();
      bool ok = false;
      for (const YamlNode& c : decl.at("choices").seq) {
        if (c.as_string() == v) ok = true;
      }
      if (!ok) fail("'" + name + "' default '" + v + "' not in choices");
      set.values_[name] = v;
    } else {
      fail("unknown param type '" + type + "' for '" + name + "'");
    }
  }
  return set;
}

int ParamSet::get_int(const std::string& name) const {
  auto it = values_.find(name);
  if (it == values_.end()) fail("unknown param '" + name + "'");
  if (!std::holds_alternative<int>(it->second)) fail("'" + name + "' is not int");
  return std::get<int>(it->second);
}

double ParamSet::get_float(const std::string& name) const {
  auto it = values_.find(name);
  if (it == values_.end()) fail("unknown param '" + name + "'");
  if (!std::holds_alternative<double>(it->second)) fail("'" + name + "' is not float");
  return std::get<double>(it->second);
}

bool ParamSet::get_bool(const std::string& name) const {
  auto it = values_.find(name);
  if (it == values_.end()) fail("unknown param '" + name + "'");
  if (!std::holds_alternative<bool>(it->second)) fail("'" + name + "' is not bool");
  return std::get<bool>(it->second);
}

std::string ParamSet::get_string(const std::string& name) const {
  auto it = values_.find(name);
  if (it == values_.end()) fail("unknown param '" + name + "'");
  if (!std::holds_alternative<std::string>(it->second)) fail("'" + name + "' is not string");
  return std::get<std::string>(it->second);
}

bool ParamSet::has(const std::string& name) const { return values_.count(name) > 0; }

}  // namespace nav_study::core

File: python/navigation/core/params.py
"""Declarative algorithm parameters loaded from configs/<category>/<algo>.yaml.

Mirrors the C++ `core/params.hpp`. The yaml *is* the declaration (per
`spec/param_schema.json`) carrying defaults; the loader validates type / range /
choices at load time so no magic numbers leak into algorithm code.
"""

from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path

import yaml

ParamValue = int | float | bool | str

_VALID_TYPES = ("int", "float", "bool", "string", "enum")


@dataclass(frozen=True)
class ParamDecl:
    name: str
    type: str
    default: ParamValue
    min: float | None = None
    max: float | None = None
    choices: list[str] | None = None
    description: str = ""


class ParamError(Exception):
    """Raised on any parameter declaration / value validation failure."""


def _check_default(decl: ParamDecl) -> ParamValue:
    """Coerce and range-check a declared default, raising ParamError on failure."""
    value = decl.default
    if decl.type == "int":
        # bool is an int subclass in Python; reject it explicitly.
        if isinstance(value, bool) or not isinstance(value, int):
            raise ParamError(f"param error: '{decl.name}' default must be int, got {value!r}")
        return _check_range(decl, float(value), value)
    if decl.type == "float":
        if isinstance(value, bool) or not isinstance(value, (int, float)):
            raise ParamError(f"param error: '{decl.name}' default must be float, got {value!r}")
        fvalue = float(value)
        return _check_range(decl, fvalue, fvalue)
    if decl.type == "bool":
        if not isinstance(value, bool):
            raise ParamError(f"param error: '{decl.name}' default must be bool, got {value!r}")
        return value
    if decl.type in ("string", "enum"):
        if not isinstance(value, str):
            raise ParamError(f"param error: '{decl.name}' default must be string, got {value!r}")
        if decl.type == "enum":
            if not decl.choices:
                raise ParamError(f"param error: enum '{decl.name}' declares no choices")
            if value not in decl.choices:
                raise ParamError(
                    f"param error: '{decl.name}' default {value!r} not in choices {decl.choices}"
                )
        return value
    raise ParamError(f"param error: '{decl.name}' has unknown type {decl.type!r}")


def _check_range(decl: ParamDecl, as_float: float, value: ParamValue) -> ParamValue:
    if decl.min is not None and as_float < decl.min:
        raise ParamError(f"param error: '{decl.name}' default {value} < min {decl.min}")
    if decl.max is not None and as_float > decl.max:
        raise ParamError(f"param error: '{decl.name}' default {value} > max {decl.max}")
    return value


class ParamSet:
    def __init__(self, algorithm: str, category: str, decls: dict[str, ParamDecl]) -> None:
        self.algorithm = algorithm
        self.category = category
        self._decls = decls
        self._values: dict[str, ParamValue] = {
            name: _check_default(decl) for name, decl in decls.items()
        }

    @classmethod
    def from_yaml(cls, path: str | Path) -> ParamSet:
        with open(path, encoding="utf-8") as fh:
            raw = yaml.safe_load(fh)
        if not isinstance(raw, dict):
            raise ParamError(f"param error: {path} is not a mapping")
        for key in ("algorithm", "category", "params"):
            if key not in raw:
                raise ParamError(f"param error: {path} missing required key '{key}'")
        params = raw["params"]
        if not isinstance(params, list):
            raise ParamError(f"param error: {path} 'params' must be a list")
        decls: dict[str, ParamDecl] = {}
        for entry in params:
            if not isinstance(entry, dict):
                raise ParamError(f"param error: {path} param entry is not a mapping")
            for req in ("name", "type", "default", "description"):
                if req not in entry:
                    raise ParamError(f"param error: {path} param missing '{req}'")
            if entry["type"] not in _VALID_TYPES:
                raise ParamError(f"param error: unknown type {entry['type']!r} for {entry['name']}")
            decl = ParamDecl(
                name=entry["name"],
                type=entry["type"],
                default=entry["default"],
                min=entry.get("min"),
                max=entry.get("max"),
                choices=entry.get("choices"),
                description=entry["description"],
            )
            decls[decl.name] = decl
        return cls(raw["algorithm"], raw["category"], decls)

    def _typed(self, name: str, expected: str) -> ParamValue:
        if name not in self._decls:
            raise ParamError(f"param error: unknown parameter '{name}'")
        decl = self._decls[name]
        if decl.type != expected:
            raise ParamError(
                f"param error: '{name}' is {decl.type}, requested as {expected}"
            )
        return self._values[name]

    def get_int(self, name: str) -> int:
        value = self._typed(name, "int")
        assert isinstance(value, int)
        return value

    def get_float(self, name: str) -> float:
        value = self._typed(name, "float")
        assert isinstance(value, (int, float))
        return float(value)

    def get_bool(self, name: str) -> bool:
        value = self._typed(name, "bool")
        assert isinstance(value, bool)
        return value

    def get_string(self, name: str) -> str:
        # 'string' and 'enum' are both surfaced as strings.
        if name not in self._decls:
            raise ParamError(f"param error: unknown parameter '{name}'")
        if self._decls[name].type not in ("string", "enum"):
            raise ParamError(f"param error: '{name}' is not a string/enum")
        value = self._values[name]
        assert isinstance(value, str)
        return value

    def has(self, name: str) -> bool:
        return name in self._decls

    def values(self) -> dict[str, ParamValue]:
        return dict(self._values)
